File: examples/py/ex_printinstancedefinitiontree.py
from scriptcontext import doc
import Rhino

def RunCommand():
  idefs = doc.InstanceDefinitions
  idefCount = idefs.Count

  if idefCount == 0:
    print "Document contains no instance definitions."
    return

  dump = Rhino.FileIO.TextLog()
  dump.IndentSize = 4

  for i in range(0, idefCount):
    DumpInstanceDefinition(idefs[i], dump, True)

  print dump.ToString()

def DumpInstanceDefinition(idef, dump, bRoot):
  if idef != None and not idef.IsDeleted:
    if bRoot:
      node = u'\u2500'
    else:
      node = u'\u2514'
    dump.Print(u"{0} Instance definition {1} = {2}\n".format(node, idef.Index, idef.Name))

    idefObjectCount = idef.ObjectCount
    if idefObjectCount  > 0:
      dump.PushIndent()
      for i in range(0, idefObjectCount):
        obj = idef.Object(i)
        if obj != None and type(obj) == Rhino.DocObjects.InstanceObject:
          DumpInstanceDefinition(obj.InstanceDefinition, dump, False) # Recursive...
        else:
          dump.Print(u"\u2514 Object {0} = {1}\n".format(i, obj.ShortDescription(False)))
      dump.PopIndent()

if __name__ == "__main__":
  RunCommand()

File: examples/cs/ex_printinstancedefinitiontree.cs
using Rhino;
using Rhino.DocObjects;
using Rhino.FileIO;
using Rhino.Commands;

namespace examples_cs
{
  [System.Runtime.InteropServices.Guid("7C98E0BC-C177-46E1-A9AE-092C63911450")]
  public class InstanceDefinitionTreeCommand : Command
  {
    public override string EnglishName { get { return "csInstanceDefinitionTree"; } }

    protected override Result RunCommand(RhinoDoc doc, RunMode mode)
    {
      var instanceDefinitions = doc.InstanceDefinitions;
      var instanceDefinitionCount = instanceDefinitions.Count;

      if (instanceDefinitionCount == 0)
      {
        RhinoApp.WriteLine("Document contains no instance definitions.");
        return Result.Nothing;
      }

      var dump = new TextLog();
      dump.IndentSize = 4;

      for (int i = 0; i < instanceDefinitionCount; i++)
        DumpInstanceDefinition(instanceDefinitions[i], ref dump, true);

      RhinoApp.WriteLine(dump.ToString());

      return Result.Success;
    }

    private void DumpInstanceDefinition(InstanceDefinition instanceDefinition, ref TextLog dump, bool isRoot)
    {
      if (instanceDefinition != null && !instanceDefinition.IsDeleted)
      {
        string node;
        node = isRoot ? "─" : "└"; // "\u2500" : "\u2514"
        dump.Print(string.Format("{0} Instance definition {1} = {2}\n", node, instanceDefinition.Index, instanceDefinition.Name));

        if (instanceDefinition.ObjectCount  > 0)
        {
          dump.PushIndent();
          for (int i = 0; i < instanceDefinition.ObjectCount ; i++)
          {
            var obj = instanceDefinition.Object(i);
            if (obj == null) continue;
            if (obj is InstanceObject)
              DumpInstanceDefinition((obj as InstanceObject).InstanceDefinition, ref dump, false); // Recursive...
            else
              dump.Print(string.Format("\u2514 Object {0} = {1}\n", i, obj.ShortDescription(false)));
          }
          dump.PopIndent();
        }
      }
    }
  }
}

